Source/WebCore/dom/Element.cpp

```diff
@@ -5363,4 +5374,40 @@ bool Element::isPopoverShowing() const
     return popoverData() && popoverData()->visibilityState() == PopoverVisibilityState::Showing;
 }
 
+// https://drafts.csswg.org/css-contain/#relevant-to-the-user
+bool Element::isRelevantToUser() const
+{
+    return hasRareData() && elementRareData()->contentRelevancyStatus();
+}
+
+OptionSet<ContentRelevancyStatus> Element::contentRelevancyStatus() const
+{
+    if (!hasRareData())
+        return { };
+    return elementRareData()->contentRelevancyStatus();
+}
+
+void Element::updateContentRelevancyStatus(ContentRelevancyStatus status, bool value)
+{
+    if (contentRelevancyStatus().contains(status) != value) {
+        ensureElementRareData().setContentRelevancyStatus(status, value);
+        invalidateStyle();
+    }
+}
+
+void Element::updateContentRelevancyStatusIfNeeded(ContentRelevancyStatus status, bool value)
+{
+    for (auto target = this; target; target = target->parentElementInComposedTree()) {
+        if (target->renderer() && target->renderer()->style().contentVisibility() == ContentVisibility::Auto) {
+            target->updateContentRelevancyStatus(status, value);
+            return;
+        }
+    }
+}
+
+void Element::contentVisibilityViewportChange(bool isIntersecting)
+{
+    updateContentRelevancyStatus(ContentRelevancyStatus::Visible, isIntersecting);
```

Comment:
Can we assert that the element has C-V auto? (Since this is called from the intersection observer, it must have C-V auto I'm guessing?)

question: Also, why aren't we iterating through the parent nodes like `updateContentRelevancyStatusIfNeeded` is doing?

Reply:
You are right, this can only be called from c-v auto elements, since only those elements add/remove an observer. The assert makes sense, added.

We are not iterating here since there is no need, we know we are dealing with the c-v: auto element so we can just update the relevancy information directly.